modules/gdscript/gdscript_cache.cpp

```diff
@@ -245,13 +339,71 @@ Error GDScriptCache::finish_compiling(const String &p_owner) {
 	return err;
 }
 
+Ref<PackedScene> GDScriptCache::get_scene(const String &p_path, const String &p_owner) {
+	MutexLock lock(singleton->mutex);
+
+	if (singleton->packed_scene_cache.has(p_path)) {
+		singleton->packed_scene_dependencies[p_path].insert(p_owner);
+		return singleton->packed_scene_cache[p_path];
+	}
+
+	Ref<PackedScene> scene;
+	scene.instantiate();
+
+	if (p_path.is_empty()) {
+		return scene;
+	}
+
+	scene->set_path(p_path);
+	singleton->packed_scene_cache[p_path] = scene;
+	singleton->packed_scene_dependencies[p_path].insert(p_owner);
+
+	scene->recreate_state();
+	scene->reload_from_file();
+	return scene;
+}
+
+void GDScriptCache::clear_unreferenced_scenes() {
+	if (singleton == nullptr) {
+		return;
+	}
+
+	MutexLock lock(singleton->mutex);
+
+	for (KeyValue<String, HashSet<String>> &E : singleton->packed_scene_dependencies) {
+		if (E.value.size() > 0 || !ResourceLoader::is_imported(E.key)) {
```

Comment:
Removed any modifications to `PackedScene` files. Added `GDScriptCache::clear_unreferenced_scenes()` that clears cached packed scenes if they are imported.

The import check is done to prevent a "cyclic load" error from `ResourceLoader`.

The way "cyclic scene references" work is pretty simple. When GDScript asks for a scene with `preload()`, a empty scene is cached first, then is reloaded with the right path. So, if another `preload()` is called within the loading of the first scene, as the empty (and loading) scene is cached, it's possible to return its reference. See [modules/gdscript/gdscript_cache.cpp:357](https://github.com/godotengine/godot/pull/67714/files#diff-a5502b90c5ce669d474f92292a3f2d6ef72ac9e1f1687534e3abce6b2640e703)

Unfortunately, I don't know how I could call `GDScriptCache::clear_unreferenced_scenes()` after the initial load.